cpu/esp32/esp-now/esp_now_netdev.c

```diff
@@ -527,9 +527,18 @@ static int _send(netdev_t *netdev, const iolist_t *iolist)
 
     /* send the packet to the peer(s) mac address */
     if (esp_now_send(_esp_now_dst, iolist->iol_base, iolist->iol_len) == 0) {
+        /**
+         * The function esp_now_send_cb does seems sporadically not to be
+         * called and _esp_now_sending remains set. This locks the entire
+         * netdev driver. The netdev driver also works without the feedback
+         * of the transmission success. In the case of a lost message, it
+         * can rely on the error control of the upper classes.
+         */
+#if 0
```

Comment:
@TimoRoth I have to come back to this locking problem :worried: I investigated it very instensively and could figure out the reason, hopefully. The problem is caused by https://github.com/RIOT-OS/RIOT/blob/842f176f94c48d245dd3fee9b64b9ebf4816af8d/cpu/esp32/esp-now/esp_now_netdev.c#L204 in https://github.com/RIOT-OS/RIOT/blob/842f176f94c48d245dd3fee9b64b9ebf4816af8d/cpu/esp32/esp-now/esp_now_netdev.c#L195

Let me try to explain the problem. We have two functions that are called from two different threads where the same mutex has to be locked:

1. `_send` function which is called from thread `net-esp-now` with pid 9 and prio 10
2. `esp_now_recv_cb` function which is called from thread `wifi`with pid 8 and prio 1

In the `_send` function, the mutex is locked with prio 10 by the `net-esp-now` thread, and `_esp_now_sending` is set to 1. Then the thread waits that `_esp_now_sending` becomes 0. `esp_now_sending` is set to 0 in the callback function `esp_now_send_cb`. This callback function is called by the thread 'wifi` with prio 1, as soon as the WLAN hardware has sent the frame.

In that case, thread `wifi` with prio 1 calls the callback function `esp_now_recv_cb` and tries to lock the mutex. Since it is already locked, thread `wifi` is blocked and placed in the waiting queue. It cannot handle events from the WiFi hardware anymore. Therefore, `esp_now_send_cb` is not called, `_esp_now_sending` keeps 1, thread `net-esp-now` keeps waiting and the mutex keeps locked. :worried: A classical deadlock :grimacing:

What are the options?
1. We don't use the mutex in `esp_now_recv_cb`
2. We have different mutexes for sending and receiving.

To decide which option is the better one, we have to figure out which parts of the `esp_now_netdev` are secured by the mutex in the receive callback function. From my point of view, please correct me if I'm wrong, this is only the ringbuffer. There are only two threads that access the ringbuffer, the `net-esp-now` thread and the `wifi` thread. Because all interrupts are disabled during the execution of the callback function and `wifi` is the thread with the highest priority, there is no way to preempt the execution of this callback function and there should be no concurrency.

I would say, this callback function has not to be secured by the mutex.

Reply:
@TimoRoth I have removed the mutex handling from `esp_now_recv_cb` and it works absolutely stable with waiting in `_send` using `_esp_now_sending` variable. I could ping 3 nodes with each other at the same time as fast as possible:
```
ping6 10000 fe80::32ae:a4ff:fe26:d1d1 52 10
```